frontend/src/DiagramEditor/DiagramEditor.jsx

```diff
@@ -55,65 +62,74 @@ export class DiagramEditor extends React.Component {
     );
   };
 
+  getSelectedConnection = () => {
+    const { selectedConnectionId } = this.state;
+
+    return this.props.data.connections.find(
+      ({ id }) => id === selectedConnectionId
+    );
+  };
+
+  onCanvasMouseMove = (e) => {};
+
   onKeyDown = (e) => {
     if (["ArrowDown", "ArrowUp", "ArrowLeft", "ArrowRight"].includes(e.key)) {
       e.preventDefault();
     }
   };
 
+  onCanvasClick = (e) => {
+    const { isConnecting } = this.state;
+
+    if (!isConnecting) {
+      this.setState({ selectedComponentId: null });
+    }
+  };
+
   onMouseUp = (e) => {
-    const { oldIsDraggingComponent, isContextMenuShowing } = this.state;
+    const {
+      isComponentContextMenuShowing,
+      isConnectionContextMenuShowing,
+      isConnecting,
+      isDraggingComponent,
+    } = this.state;
 
     this.setState({
       isDraggingComponent: false,
       isPanning: false,
+      isConnecting: false,
     });
-    if (isContextMenuShowing) {
+
+    if (isComponentContextMenuShowing) {
       this.setState({
-        isContextMenuShowing: false,
-        selectedComponentId: null,
+        isComponentContextMenuShowing: false,
       });
       return;
     }
-
-    if (!oldIsDraggingComponent) {
+    if (isConnectionContextMenuShowing) {
+      this.setState({
+        isConnectionContextMenuShowing: false,
+      });
       return;
     }
 
-    const wholeDeltaX = e.clientX - (this.state.initialMouseX || 0);
-    const wholeDeltaY = e.clientY - (this.state.initialMouseY || 0);
-
-    if (!wholeDeltaX && !wholeDeltaY) {
-      console.log("we have no delta");
-      return;
+    if (!isConnecting) {
+      this.setState({
+        selectedComponentId: null,
```

Comment:
It looks like `onMouseUp` gets triggered when `onComponentMouseUp` is triggered. Here we're setting `selectedComponentId` to `null`, whereas in `onComponentMouseUp` we're setting it to the id of the component. Isn't there a bit of a competition here as to whomever goes last wins? Or am I just confused?

Reply:
I've eliminated that condition by separating the mouse up handlers into 2 parts: component, canvas and window. The component handler stops the propagation of the event, so it never gets to the other 2 handlers if you click on a component. Otherwise, the other 2 run, so there is no race condition. The canvas and window handlers also do different things, so they don't compete with one another

Reply:
Great!